### ISSUE TITLE
Tuist still generates targets for incompatible dependencies

### ISSUE THREAD

**Issue by Mstrodl** [tuist/tuist#5295]
### What happened?

When generating a project with a swift package manager dependency containing a conditional package like this for another platform:
```
                .product(name: "Datadog", package: "dd-sdk-ios", condition: .when(platforms: [.iOS])),
                .product(name: "DatadogCrashReporting", package: "dd-sdk-ios", condition: .when(platforms: [.iOS])),
```
Tuist will still include a `datadog_watchos` dependency in the tree (and all transitive dependencies obviously). In our case, Datadog's SDK won't compile for watchOS, so this poses a pretty big issue for us, meaning our project won't compile!
If I manually edit the graph.json and remove the `datadog_watchos` (and related) entries, I can `tuist generate` and compile the project just fine, so this appears to be something tuist is doing that it shouldn't be.

### How do we reproduce it?

1. Use this demo project: https://github.com/Mstrodl/watchos-dependency-demo
2. Run `tuist fetch`
3. Then `tuist generate`
4. Build the watch app
5. Note the presence of the `datadog_watchos` dependency

### Error log

![image](https://github.com/tuist/tuist/assets/6877780/4361399b-874d-4490-b70c-29aedd8db899)
Note the `Datadog_watchos` dependencies

### macOS version

13.4.1

### Tuist version

3.21.0

### Xcode version

Version 15.0 beta 3 (15A5195k)

**Comment by danyf90** [comment]
Hello 👋 
Could you please clarify why `watchosapp` depends on `ExternalDeps` if they are not for watchos?

**Comment by Mstrodl** [comment]
> Could you please clarify why `watchosapp` depends on `ExternalDeps` if they are not for watchos?
Hi! `ExternalDeps` is for watchos _and_ ios. On iOS it is enhanced by the presence of Datadog. I didn't add any actual content to `ExternalDeps` since it seemed out of scope for the sample project. In our app, we have a private dependency with all of our API models and other shared business logic that also handles setting up some user information on Datadog (fenced off with `#if os(iOS)`). Let me know if you have any other questions!

**Comment by danyf90** [comment]
I see 👍
I think we just need to fix a bit the PackageInfoMapper to support the when condition then 👍

**Comment by waltflanagan** [comment]
This should be resolved when #5595 is merged.

**Comment by waltflanagan** [comment]
@Mstrodl Looks like there was a 2nd bug behind the first one that broke the parsing of the condition.  the logic for applying it when it's non-nil works great, but no if it's always nil.  

Once #5846 is merged it will go out in the next release.
